### PR TITLE
Task 2.3: Filter students by search and sections

### PR DESCRIPTION
**Pull request by abeeto**
This branch is based on changes introduced in PR #37 
- retrieves new set of section using `fetch` on a new endpoint configured in https://github.com/Teamable-Analytics/teamable-backend/pull/37
- uses query param `sections` to get list of sections to filter on.
  - e.g. `?sections=1.5.12` (chaining the id of sections using '.'
- new file `students/(table)/student-table-filter.tsx`:  
  - was created so that instead of using react-table's client side filtering, it just updates states for serverside api fetching.
  - based on `data-table-faceted-filter.tsx`: honestly not even sure if that was an original ShadCn component need to check.

UPDATE MAR 29:  
- IMPORTANT: 'filter students by section' is not fully functional. Setting the query params in the browser works to filter the students, but the  student sections filter UI does not work. Will be a separate issue/PR.

### CODE REVIEW

File: src/app/students/(hooks)/useStudents.tsx [hunk 1]
```diff
@@ -12,54 +21,111 @@ type StudentsContextType = {
...
+    const searchParams = new URLSearchParams()
+    Object.entries(params).forEach(([key, value]) => {
+        if (value) {
+            searchParams.set(key, value.toString())
+        }
+    })
+    return searchParams.toString()
 }
-const StudentsContext = createContext<StudentsContextType | undefined>(undefined)
+// todo
```
Comment (Opeyem1a) [inline src/app/students/(hooks)/useStudents.tsx:50]: nit;

remove

File: src/app/students/(hooks)/useStudents.tsx [hunk 2]
```diff
@@ -12,54 +21,111 @@ type StudentsContextType = {
...
         const fetchStudents = async () => {
-            const baseURL = process.env.NEXT_PUBLIC_BACKEND_URL
-            const courseMemberData = await fetch(`${baseURL}/api/v1/course-members/course/${fixedCourseNum}/?page=${pageIndex}&per_page=${pageSize}`).then(res => res.json())
+            const courseMemberResponse = await fetch(`${process.env.NEXT_PUBLIC_BACKEND_URL}/api/v1/course-members/course/${FIXED_COURSE_NUM}/?${queryString}`)
+            const courseMemberData = await courseMemberResponse.json()
             const studentsToDisplay: Student[] = courseMemberData.results.map((member: any) => ({
-                id: member.id,
-                name: member.user.last_name + ',' + member.user.first_name,
+                id: member.user.id,
+                name: `${member.user.last_name}, ${member.user.first_name}`,
```
Comment (Opeyem1a) [inline src/app/students/(hooks)/useStudents.tsx:92]: nit; (not for this PR)

now that I think about it, probably unwise to concatenate this here, only to display them in separate columns anyway yknow

File: src/app/students/(table)/student-table-filter.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import {
+    Popover,
+    PopoverContent,
+    PopoverTrigger,
+} from "@/components/ui/popover"
+import { Separator } from "@/components/ui/separator"
+import { useStudents } from "../(hooks)/useStudents"
+import { useState } from "react"
+
+type DataTableFacetedFilterProps<TData, TValue> = {
```
Comment (Opeyem1a) [inline src/app/students/(table)/student-table-filter.tsx:25]: nit;

this type has the "it is defined in another file" vibe, no need to redefine it here if that is the case

**Review by Opeyem1a — APPROVED** [review]

File: src/app/students/(hooks)/useStudents.tsx (hunk 2 above)
Comment (abeeto) [reply src/app/students/(hooks)/useStudents.tsx:91]: will make an issue for it

**Pull request merged**